vault/ha.go

```diff
@@ -576,7 +576,9 @@ func (c *Core) waitForLeadership(newLeaderCh chan func(), manualStepDownCh, stop
 					c.logger.Error("clearing leader advertisement failed", "error", err)
 				}
 
-				c.heldHALock.Unlock()
+				if err := c.heldHALock.Unlock(); err != nil {
+					c.logger.Error("unlocking HA lock failed", "error", err)
```

Comment:
After some discussion, we decided that this line is OK to leave in.  But please add a check to Unlock() (which calls your modified Delete) to catch the ConditionCheck error and not report it as an error, similar to that handling here:  https://github.com/hashicorp/vault/blob/54d8c25a9e225dc0e99c740524eab602b07e6839/physical/dynamodb/dynamodb.go#L632

Reply:
I do think this case is a bit different from the tryToLock one because in that case it is speculatively attempting to update and has the expectation that it can reasonably fail.  Whereas for Unlock() this shouldn't really expected unless something aberrant has happened. 

Though as long as Vault will always attempt unlock when it steps down, even if it knows there is a new leader already, it's reasonable to not report the error.

Reply:
It is interesting b/c I've seen this occur in Unlock at least once, but that was during a simulated network outage. Much more common (and I didn't mention this previously) is `[ERROR] storage.dynamodb: error renewing leadership lock: error="ConditionalCheckFailedException: ...`.  This occurs with your current build during `operator step-down`. So a couple of things:

1. This error catch should be just in dynamodb.go, not ha.go which is backend agnostic. e.g. move to after you call `DeleteItem()` in the dynamo implementation of Unlock().
2. We also need to catch the one triggering the "renewing leadership lock" message. Both tryToLock and periodicallyRenewLock rely updateItem, so I think centralizing the check in updateItem is reasonable. i.e. if the error is the ConditionCheck, just make it nil instead.
3. (Optional) A filterError() method that just does the transformation of err->nil if ConditionCheck (or a no-op if no) might benefit 1 and 2, and keeps it clean if other such benign errors ever are added.

I appreciate the effort on this! I'd like to get this merged fairly soon, so if you need any help I'm able to push to your branch as well.

Reply:
Those all make sense to me and should be taken care of now.

Reply:
Alas, my suggested refactor was too much. The presence of an error is needed from `updateItem()` for correct `tryToLock` behavior. I've changed this slightly to restore that, and now it seems to work fine. There is no flapping of the active node, spurious errors aren't showing up in the logs, a situations like the lock file being deleted or network being lost are handled OK. Let me know what you think.

Reply:
Changes look good to me.